File: 2024/01.swift
#!/usr/bin/env swift

import Foundation
import RegexBuilder
import Testing

let start = Date()
var input = try String(contentsOfFile: #file.replacingOccurrences(of: ".swift", with: ".input"))

var lhs: [Int] = []
var rhs: [Int] = []

for line in input.split(separator: "\n") {
    let matches = line.matches(of: OneOrMore(.digit))
    lhs.append(Int(matches[0].0)!)
    rhs.append(Int(matches[1].0)!)
}

lhs.sort()
rhs.sort()

assert(lhs.count == rhs.count)

var distance = 0
for (lhs, rhs) in zip(lhs, rhs) {
    distance += abs(lhs - rhs)
}

print("part1", distance)

/// int: count
var rhsCount = [Int: Int]()
for rhs in rhs {
    rhsCount[rhs, default: 0] += 1
}

var similarity = 0
for lhs in lhs {
    similarity += lhs * rhsCount[lhs, default: 0]
}

print("part2", similarity)
print("elapsed time in seconds:", -start.timeIntervalSinceNow)

File: 2024/01.r
#!/usr/bin/env Rscript

start = Sys.time()
library("this.path")
input = read.table(gsub('r$', 'input', this.path()), header = FALSE)

result1 = sum(abs(sort(input[,1]) - sort(input[,2])))

result2 = 0
for (lhs in input[,1]) {
    result2 = result2 + lhs * length(input[,2][input[,2] == lhs])
}

paste("part1", result1, result1 == 1258579)
paste("part2", result2, result2 == 23981443)
paste("elapsed time in seconds:", Sys.time() - start)
